src/mlpack/core/cv/purged_k_fold_cv.hpp

```diff
@@ -0,0 +1,282 @@
+/**
+ * @file core/cv/purged_k_fold_cv.hpp
+ * @author Felix Patschkowski
+ *
+ * purged k-fold cross-validation.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_CV_PURGED_K_FOLD_CV_HPP
+#define MLPACK_CORE_CV_PURGED_K_FOLD_CV_HPP
+
+#include "k_fold_cv_base.hpp"
+
+namespace mlpack {
+
+/**
+ * A purged k-fold CV is suitable for time series with informational
+ * overlap from the training set into the validation set and vice
+ * versa. This CV will remove items from the training set that overlap
+ * with the validation set.
+ */
+template<typename MLAlgorithm,
+         typename Metric,
+         typename MatType = arma::mat,
+         typename PredictionsType =
+             typename MetaInfoExtractor<MLAlgorithm, MatType>::PredictionsType,
+         typename WeightsType =
+             typename MetaInfoExtractor<MLAlgorithm, MatType,
+                 PredictionsType>::WeightsType>
+class PurgedKFoldCV :
+    private KFoldCVBase<
+        PurgedKFoldCV<MLAlgorithm, Metric, MatType, PredictionsType, WeightsType>,
+        MLAlgorithm,
+        Metric,
+        MatType,
+        PredictionsType,
+        WeightsType>
+{
+ public:
+  /**
+    * This constructor can be used for regression algorithms and for binary
+    * classification algorithms.
+    *
+    * @param k Number of folds (should be at least 2).
+    * @param[in] embargoPercentage Percentage of the dataset size that is used
+    *                              to extend the duration of the intervals in
+    *                              the validation set to remove elements from
+    *                              the training set after the validation set
+    *                              (embargo). 
+    * @param xs Data points to cross-validate on.
+    * @param ys Predictions (labels for classification algorithms and responses
+    *     for regression algorithms) for each data point.
+    * @param[in] intervals A `2 x m` matrix, where each column stores the start
+    *                      and end sample of an interval.
+    */
+  PurgedKFoldCV(
+      size_t k,
+      double embargoPercentage,
+      const MatType& xs,
+      const PredictionsType& ys,
+      const MatType& intervals);

```

Comment:
Also, is it always necessary to pass the `intervals` matrix?  Or does it make sense to have a default argument (or similar) such that when `intervals` is not specified, we assume that column `i` of `xs` corresponds only to time step `i`---so there is no overlap in samples.

That would match the simple case in the image given with #3830, where `xs` looks like stock prices (or something like this) and each sample in `xs` is just the price at one time step.  In that case it seems like you would still want to do purged k-fold CV, but the `intervals` matrix would just be of the form, e.g.,

```
0, 0
1, 1
2, 2
3, 3
...
```

To me it seems like this would be the most common case to use `PurgedKFoldCV`, so having a 'default' for `intervals` could be useful.  What do you think?

Reply:
We now already have 6 constructors. I think defaulting that argument will introduce another 6 sets of constructors so I'd go for the current solution. The user is free to supply their `intervals` matrix as you suggested at any time.

Reply:
I still have my concerns about this design. In case we use a defaulted `intervals`, then one could also use that standard `KFoldVC` instead of the purged version.